### ISSUE TITLE
Dynamic image stream for form fill partials via get requests

### ISSUE THREAD

**Issue by karpet** [BlueprintKansas/ksvotes.org#1]
<a href="https://github.com/jakelowen"><img src="https://avatars0.githubusercontent.com/u/5368008?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [jakelowen](https://github.com/jakelowen)**
_Monday Feb 05, 2018 at 22:36 GMT_
_Originally opened as https://github.com/gpsimpact/open_voter_reg/issues/1_

----

Objective:  In addition to full form renders, let's allow the user of dynamic partial form renders.  I.e. a png that is just section 1 (first line) of the federal voter reg form. We can have a express get request handler that looks for url query string paramters and composes and image on the fly. A request might look like:

"whatever.com/image-generator?background_image=http://...&drawText=%5B%7B%22x%22%3A20%2C%22y%22%3A20%2C%22text%22%3A%22foobar%22%7D%2C%7B%22x%22%3A50%2C%22y%22%3A50%2C%22text%22%3A%22pickles%22%7D%5D"

the drawText portion could be generated with:
```
var querystring = require('querystring');
var text = [
  { x: 20, y: 20, text: "foobar" },
  { x: 50, y: 50, text: "pickles" },
]
querystring.stringify({ 'drawText': JSON.stringify(text) })
```

So basically pas in a background image and json array of x,y,text and dynamically render the image via streams

**Comment by karpet** [comment]
<a href="https://github.com/jakelowen"><img src="https://avatars0.githubusercontent.com/u/5368008?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [jakelowen](https://github.com/jakelowen)**
_Monday Feb 05, 2018 at 22:50 GMT_

----

Here is a different snippet from my scratchpad that shows how to dynamically render and stream images via graphicksmagick / express.. The idea below is slightly different as it pulls the data from a hard coded dict instead of url query string parameters but the gist of how to work with image streams is here:

```
import express from "express";
import gm from "gm"
import request from "request"
import { map } from "lodash"

const goat = "http://cdn.playbuzz.com/cdn/c1ffedc1-2f64-4503-8c91-689bb8c48218/8325ab64-99f2-4ce1-9da3-c98b8ae7e395.jpg"


const app = express()

const ref_data = {
  alpha: {
    foo: "thisisfooalpha",
    bar: "thisisbaralpha",
    baz: "thisisbazalpha"
  },
  beta: {
    foo: "thisisfoobeta",
    bar: "thisisbarbeta",
    baz: "thisisbazbeta"
  }
}

const ref_templates = {
  goat: {
    fields: {
      foo: { x: 20, y: 30 },
      bar: { x: 30, y: 40 },
      baz: { x: 40, y: 50 }
    },
    base: {
      dimensions: { x: 800, y: 600 },
      source: "http://cdn.playbuzz.com/cdn/c1ffedc1-2f64-4503-8c91-689bb8c48218/8325ab64-99f2-4ce1-9da3-c98b8ae7e395.jpg"
    },
    style: {
      font: "Roboto-Regular.ttf",
      size: 20
    }
  }
}

app.get('/generate', function (req, res) {
  res.set('Content-Type', 'image/png');
  const template = ref_templates[req.query.template]
  const data = ref_data[req.query.data]
  const imageStream = gm(template.base.source)
  console.log(template)
  if (template.style && template.style.font && template.style.size) {
    console.log(template.style.font, template.style.size)
    imageStream.font(template.style.font, template.style.size)
  }
  map(data, (value, key) => {
    imageStream.drawText(template.fields[key].x, template.fields[key].y, value)
  })
  imageStream
    .setFormat('png')
    .stream()
    .pipe(res)
});

app.get('/', (req, res) => res.send('Hello World!'))
app.listen(3000, '0.0.0.0', () => console.log('Example app listening on port 3000!'))
```
